File: shared/src/solana/builders/register-agent.ts
import { PublicKey, SystemProgram, VersionedTransaction } from "@solana/web3.js";
import { Program } from "@coral-xyz/anchor";
import type { Basira } from "../idl/basira";
import { agentPda } from "../pdas";
import { buildVersionedTx } from "./_tx";

/**
 * Build a `register_agent` tx signed by the platform authority (NOT the agent).
 * Agents stay off-chain — the platform's keeper keypair is the sole on-chain
 * signer and fee payer. The agent wallet is passed as an instruction argument
 * and used for the PDA seed.
 */
export async function buildRegisterAgentTx({
  agentWallet,
  platformAuthority,
  recentBlockhash,
  program,
}: {
  agentWallet: PublicKey;
  platformAuthority: PublicKey;
  recentBlockhash: string;
  program: Program<Basira>;
}): Promise<{ tx: VersionedTransaction; agentAccount: PublicKey }> {
  const [agentAccount] = agentPda(agentWallet);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const ix = await (program.methods as any)
    .registerAgent(agentWallet)
    .accounts({
      platformAuthority,
      agentAccount,
      systemProgram: SystemProgram.programId,
    })
    .instruction();
  return {
    tx: buildVersionedTx(platformAuthority, recentBlockhash, [ix]),
    agentAccount,
  };
}

File: program/programs/basira/src/instructions/register_agent.rs
use anchor_lang::prelude::*;

use crate::constants::{AGENT_SEED, PLATFORM_AUTHORITY};
use crate::errors::BasiraError;
use crate::state::{AgentAccount, AgentStatus};

/// `register_agent` no longer requires the agent to sign. Agents are off-chain
/// entities; making them custody a Solana keypair just to create their
/// reputation PDA was a barrier to onboarding. The platform authority — the
/// same keypair the daemon uses for auto-release sweeps and submit_deliverable
/// — signs and pays on the agent's behalf. The agent wallet is passed as an
/// argument and used both for the PDA seed and as `agent.wallet`.
#[derive(Accounts)]
#[instruction(agent_wallet: Pubkey)]
pub struct RegisterAgent<'info> {
    #[account(mut, address = PLATFORM_AUTHORITY @ BasiraError::NotPlatformAuthority)]
    pub platform_authority: Signer<'info>,

    #[account(
        init,
        payer = platform_authority,
        space = 8 + AgentAccount::INIT_SPACE,
        seeds = [AGENT_SEED, agent_wallet.as_ref()],
        bump,
    )]
    pub agent_account: Account<'info, AgentAccount>,

    pub system_program: Program<'info, System>,
}

pub fn register_agent_handler(ctx: Context<RegisterAgent>, agent_wallet: Pubkey) -> Result<()> {
    let agent = &mut ctx.accounts.agent_account;
    let clock = Clock::get()?;

    agent.wallet = agent_wallet;
    agent.registered_at = clock.unix_timestamp;
    agent.completed_count = 0;
    agent.disputed_count = 0;
    agent.status = AgentStatus::Active;
    agent.bump = ctx.bumps.agent_account;

    Ok(())
}
